### PR TITLE
Create "add application" odo  pipeline sub-comment

### PR DESCRIPTION
**Pull request by gaganhegde**
**What type of PR is this?**
https://issues.redhat.com/projects/GITOPS/issues/GITOPS-70

**What does does this PR do / why we need it**:
`odo pipelines service add --gitops-repo <username>/<repo> --app-name <application name> --output <output folder>  --env-name <environment name> --service-git-repo <username>/<repo>i  --service-webhook-secret <secret>`

├── applications
│       └── <application name>
│       ├── base
│       │   └── kustomization.yaml
│       ├── kustomization.yaml
│       └── overlays
│           └── kustomization.yaml
├── environments
│   ├── cicd
│   │   ├── base
│   │   │   ├── kustomization.yaml
│   │   │   └── pipelines
│   │   │       ├── 01-namespaces
│   │   │       │   └── cicd-environment.yaml
│   │   │       ├── 02-rolebindings
│   │   │       │   ├── pipeline-service-account.yaml
│   │   │       │   ├── pipeline-service-rolebinding.yaml
│   │   │       │   └── pipeline-service-role.yaml
│   │   │       ├── 03-secrets
│   │   │       │   ├── docker-config.yaml
│   │   │       │   └── gitops-webhook-secret.yaml
│   │   │       ├── 04-tasks
│   │   │       │   ├── deploy-from-source-task.yaml
│   │   │       │   └── deploy-using-kubectl-task.yaml
│   │   │       ├── 05-pipelines
│   │   │       │   ├── app-cd-pipeline.yaml
│   │   │       │   ├── app-ci-pipeline.yaml
│   │   │       │   ├── cd-deploy-from-push-pipeline.yaml
│   │   │       │   └── ci-dryrun-from-pr-pipeline.yaml
│   │   │       ├── 06-bindings
│   │   │       │   ├── github-pr-binding.yaml
│   │   │       │   └── github-push-binding.yaml
│   │   │       ├── 07-templates
│   │   │       │   ├── app-cd-build-pr-template.yaml
│   │   │       │   ├── app-ci-build-pr-template.yaml
│   │   │       │   ├── cd-deploy-from-push-template.yaml
│   │   │       │   └── ci-dryrun-from-pr-template.yaml
│   │   │       ├── 08-eventlisteners
│   │   │       │   └── cicd-event-listener.yaml
│   │   │       ├── 09-routes
│   │   │       │   └── gitops-webhook-event-listener.yaml
│   │   │       └── kustomization.yaml
│   │   └── overlays
│   │       └── kustomization.yaml
│   └── <environment name>
│       ├── base
│       │   ├── kustomization.yaml
│       │   ├── namespace.yaml
│       │   └── rolebinding.yaml
│       └── overlays
│           └── kustomization.yaml
├── manifest.yaml
└── services
    └── <service name>
        ├── base
        │   ├── config
        │   │   ├── app-webhook-secret.yaml
        │   │   └── kustomization.yaml
        │   └── kustomization.yaml
        └── overlays
            ├── eventlistener_patch.yaml
            ├── kustomization.yaml
            └── pipeline_patch.yaml

### CODE REVIEW

File: pkg/odo/cli/cli.go [hunk 1]
```diff
@@ -175,6 +175,7 @@ func NewCmdOdo(name, fullName string) *cobra.Command {
 		debug.NewCmdDebug(debug.RecommendedCommandName, util.GetFullName(fullName, debug.RecommendedCommandName)),
 		pipelines.NewCmdComponent(pipelines.RecommendedCommandName, util.GetFullName(fullName, pipelines.RecommendedCommandName)),
 		pipelines.NewCmdInit(pipelines.InitRecommendedCommandName, util.GetFullName(fullName, pipelines.InitRecommendedCommandName)),
+		pipelines.NewCmdAdd(pipelines.AddRecommendedCommandName, util.GetFullName(fullName, pipelines.AddRecommendedCommandName)),
 	)
```
Comment (wtam2018) [inline pkg/odo/cli/cli.go]: Take a look at for example preference.NewCmdPreference above for how to create sub-command.   You want to create a pipelines.NewCmdService().  In pipelines.NewCmdService(), I would add NewCmdAdd().   The command line interface shoould be "odo pipelines services add"

File: pkg/odo/cli/pipelines/add.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
+package pipelines
```
Comment (wtam2018) [inline pkg/odo/cli/pipelines/add.go]: This file collides with GITOPS-74

File: pkg/odo/cli/pipelines/add.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+	"github.com/openshift/odo/pkg/odo/genericclioptions"
+	"github.com/openshift/odo/pkg/pipelines"
+	"github.com/spf13/cobra"
+
+	ktemplates "k8s.io/kubernetes/pkg/kubectl/util/templates"
+)
+
+const (
+	// InitRecommendedCommandName the recommended command name
```
Comment (wtam2018) [inline pkg/odo/cli/pipelines/add.go]: Pls fix comment.

File: pkg/odo/cli/pipelines/init.go [hunk 4]
```diff
@@ -28,11 +28,14 @@ var (
...
+	gitOpsRepo               string // repo to store Gitops resources e.g. org/repo
+	gitOpsWebhookSecret      string // used to create Github's shared webhook secret for gitops repo
+	output                   string // path to add Gitops resources
+	prefix                   string // used to generate the environments in a shared cluster
+	skipChecks               bool   // skip Tekton installation checks
+	imageRepo                string
+	internalRegistryHostname string
 	// generic context options common to all commands
 	*genericclioptions.Context
 }
```
Comment (wtam2018) [inline pkg/odo/cli/pipelines/init.go:42]: Why these changes?  Bad merge?

File: pkg/odo/cli/pipelines/init.go [hunk 5]
```diff
@@ -65,11 +68,14 @@ func (io *InitParameters) Validate() error {
...
-		SkipChecks:          io.skipChecks,
+		DeploymentPath:           io.deploymentPath,
+		GitOpsWebhookSecret:      io.gitOpsWebhookSecret,
+		GitOpsRepo:               io.gitOpsRepo,
+		Output:                   io.output,
+		Prefix:                   io.prefix,
+		ImageRepo:                io.imageRepo,
+		SkipChecks:               io.skipChecks,
+		InternalRegistryHostname: io.internalRegistryHostname,
 	}
```
Comment (wtam2018) [inline pkg/odo/cli/pipelines/init.go:82]: Bad merge?

File: pkg/odo/cli/pipelines/init.go [hunk 6]
```diff
@@ -96,6 +102,10 @@ func NewCmdInit(name, fullName string) *cobra.Command {
 	initCmd.Flags().StringVar(&o.output, "output", ".", "folder path to add Gitops resources")
 	initCmd.Flags().StringVarP(&o.prefix, "prefix", "p", "", "add a prefix to the environment names")
 	initCmd.Flags().BoolVarP(&o.skipChecks, "skip-checks", "b", false, "skip Tekton installation checks")
+	initCmd.Flags().StringVar(&o.imageRepo, "image-repo", "", "image repository in this form <registry>/<username>/<repository> or <project>/<app> for internal registry")
+	initCmd.MarkFlagRequired("image-repo")
+	initCmd.Flags().StringVar(&o.deploymentPath, "deployment-path", "deploy", "deployment folder path name")
+	initCmd.Flags().StringVar(&o.internalRegistryHostname, "internal-registry-hostname", "image-registry.openshift-image-registry.svc:5000", "internal image registry hostname")
```
Comment (wtam2018) [inline pkg/odo/cli/pipelines/init.go:114]: Bad merge?

File: pkg/pipelines/add.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+	ssv1alpha1 "github.com/bitnami-labs/sealed-secrets/pkg/apis/sealed-secrets/v1alpha1"
+	"github.com/mitchellh/go-homedir"
+	"github.com/openshift/odo/pkg/pipelines/eventlisteners"
+	"github.com/openshift/odo/pkg/pipelines/meta"
+	"github.com/openshift/odo/pkg/pipelines/roles"
+	"github.com/openshift/odo/pkg/pipelines/secrets"
+	triggersv1 "github.com/tektoncd/triggers/pkg/apis/triggers/v1alpha1"
+)
+
+// InitParameters is a struct that provides flags for initialise command
```
Comment (wtam2018) [inline pkg/pipelines/add.go]: Please fix all these comments.

File: pkg/pipelines/init.go [hunk 8]
```diff
@@ -20,11 +20,14 @@ import (
```
Comment (wtam2018) [inline pkg/pipelines/init.go:16]: I don't think you are changing the "init" command, right?

Comment (DhanashriPalodkar) [reply pkg/pipelines/init.go:20]: No we are not, I guess it's a bad merge.. will check it

File: pkg/pipelines/add_application.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+	if !exists {
+		return fmt.Errorf("%s does not exist at %s", gitopsName, gitopsPath)
+	}
+	configPath := filepath.Join(gitopsPath, servicesDir, ServiceRepo)
+
+	files := createResourcesConfig(outputs, o.EnvName, o.DockerCfgJson)
+
+	fileNames, err := writeResources(configPath, files)
+
+	fmt.Println("These are the fileNames", fileNames)
```
Comment (wtam2018) [inline pkg/pipelines/add_application.go]: Please remove.

File: pkg/pipelines/add_application.go [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+		return err
+	}
+	if err := addKustomize("bases", []string{"../base"}, filepath.Join(gitopsPath, servicesDir, ServiceRepo, baseDir, kustomize)); err != nil {
+		return err
+	}
+	if err := addKustomize("bases", []string{"../base"}, filepath.Join(gitopsPath, servicesDir, ServiceRepo, overlaysDir, kustomize)); err != nil {
+		return err
+	}
+	kustomize1 := map[string][]string{
+		"bases":     []string{"../../../../envs/chetan-cicd-environment/overlays/"},
```
Comment (wtam2018) [inline pkg/pipelines/add_application.go]: Obviously, fix the hardcoded path

Comment (bigkevmcd) [reply pkg/pipelines/add_application.go]: You'll want to use `filepath.Rel` for this https://golang.org/pkg/path/filepath/#Rel

File: pkg/odo/cli/cli.go (hunk 1 above)
Comment (ishitasequeira) [reply pkg/odo/cli/cli.go]: changes have been made

File: pkg/pipelines/init.go (hunk 8 above)
Comment (ishitasequeira) [reply pkg/pipelines/init.go:16]: pipelines and templates need to be added

File: pkg/manifest/init.go [hunk 11]
```diff
@@ -141,28 +158,70 @@ func CreateResources(prefix, gitOpsRepo, gitOpsWebhook string) (map[string]inter
...
+		// add secret and sa to outputs
+		outputs[serviceAccountPath] = roles.AddSecretToSA(sa, dockerSecretName)
+	}
+
 	outputs[rolebindingsPath] = roles.CreateRoleBinding(meta.NamespacedName(cicdNamespace, roleBindingName), sa, "ClusterRole", roles.ClusterRoleName)
 
 	outputs[tasksPath] = tasks.CreateDeployFromSourceTask(cicdNamespace, GetPipelinesDir("", prefix))
 	outputs[ciPipelinesPath] = pipelines.CreateCIPipeline(meta.NamespacedName(cicdNamespace, "ci-dryrun-from-pr-pipeline"), cicdNamespace)
 	outputs[cdPipelinesPath] = pipelines.CreateCDPipeline(meta.NamespacedName(cicdNamespace, "cd-deploy-from-push-pipeline"), cicdNamespace)
+	outputs[appCiPipelinesPath] = pipelines.CreateCIPipeline(meta.NamespacedName(cicdNamespace, "app-ci-pipeline"), cicdNamespace)
```
Comment (chetan-rns) [inline pkg/manifest/init.go:178]: Pipelines are repeated here. We need to generate separate pipelines that'll deploy the app to a given environment

Comment (chetan-rns) [inline pkg/manifest/init.go]: Pipelines are repeated here. We need to generate separate pipelines that'll deploy the app to a given environment

File: pkg/odo/cli/pipelines/service/add_application.go [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+		Use:     name,
+		Short:   AddShortDesc,
+		Long:    AddLongDesc,
+		Example: fmt.Sprintf(AddExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			genericclioptions.GenericRun(o, cmd, args)
+		},
+	}
+
+	addCmd.Flags().StringVar(&o.output, "output", "", "file path to output folder")
```
Comment (chetan-rns) [inline pkg/odo/cli/pipelines/service/add_application.go:100]: output flag should be made mandatory

File: pkg/pipelines/add_application.go [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+	createPatchFiles(outputs, o.ServiceGitRepo)
+
+	CreatePatchKustomiseFile(outputs, filepath.Join(overlaysDir, manifest.Kustomize))
+
+	environmentName := manifest.NamespaceNames(o.Prefix)
+
+	createKustomizeMod(outputs, kustomizeModPath, environmentName["cicd"])
+
+	secretName := fmt.Sprintf("svc-%s-secret", ServiceRepo)
+	files := createResourcesConfig(outputs, o.ServiceWebhookSecret, environmentName["cicd"], secretName)
```
Comment (chetan-rns) [inline pkg/pipelines/add_application.go]: These resources are pointing to the wrong namespace. They should be pointing to `o.EnvName`

File: pkg/pipelines/add_application.go [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+
+	return outputs
+}
+
+func createPatchFiles(outputs map[string]interface{}, serviceRepo string) {
+	t := []patchStringValue{
+		{
+			Op:    "add",
+			Path:  "/spec/triggers/-",
+			Value: eventlisteners.CreateListenerTrigger("app-ci-build-from-pr", eventlisteners.StageCIDryRunFilters, serviceRepo, "github-pr-binding", "app-ci-template"),
```
Comment (chetan-rns) [inline pkg/pipelines/add_application.go]: `createGitHubInterceptor()` should accept parameters for webhook secrets. Currently it always point to `gitops-webhook-secret`

File: pkg/odo/cli/pipelines/pipelines.go [hunk 15]
```diff
@@ -14,7 +16,7 @@ const RecommendedCommandName = "pipelines"
 func NewCmdPipelines(name, fullName string) *cobra.Command {
 
 	initCmd := NewCmdInit(InitRecommendedCommandName, odoutil.GetFullName(fullName, InitRecommendedCommandName))
-
+	ServiceCmd := service.NewCmdAddService(service.ServiceRecommendedCommandName, odoutil.GetFullName(fullName, service.ServiceRecommendedCommandName))
```
Comment (chetan-rns) [inline pkg/odo/cli/pipelines/pipelines.go]: `service.NewCmdService()` should be used instead of `service.NewCmdAddService()`

File: pkg/manifest/init.go [hunk 16]
```diff
@@ -141,28 +158,70 @@ func CreateResources(prefix, gitOpsRepo, gitOpsWebhook string) (map[string]inter
...
+	outputs[appCdPipelinesPath] = pipelines.CreateCDPipeline(meta.NamespacedName(cicdNamespace, "app-cd-pipeline"), cicdNamespace)
 	outputs[prBindingPath] = triggers.CreatePRBinding(cicdNamespace)
 	outputs[pushBindingPath] = triggers.CreatePushBinding(cicdNamespace)
 	outputs[prTemplatePath] = triggers.CreateCIDryRunTemplate(cicdNamespace, saName)
+	outputs[appCIBuildPRTemplatePath] = triggers.CreateDevCIBuildPRTemplate(cicdNamespace, saName, imageRepo)
```
Comment (chetan-rns) [inline pkg/manifest/init.go:187]: PipelineRuns in `CreateDevCIBuildPRTemplate()` and `CreateDevCDDeployTemplate()` are not pointing to the right pipelines

File: pkg/odo/cli/pipelines/init.go (hunk 6 above)
Comment (ishitasequeira) [reply pkg/odo/cli/pipelines/init.go:114]: made the required changes

File: pkg/odo/cli/pipelines/init.go (hunk 5 above)
Comment (ishitasequeira) [reply pkg/odo/cli/pipelines/init.go:82]: made the required changes

File: pkg/odo/cli/pipelines/init.go (hunk 4 above)
Comment (ishitasequeira) [reply pkg/odo/cli/pipelines/init.go:42]: made the required changes

File: pkg/odo/cli/pipelines/pipelines.go (hunk 15 above)
Comment (ishitasequeira) [reply pkg/odo/cli/pipelines/pipelines.go]: resolved

File: pkg/pipelines/add_application.go (hunk 14 above)
Comment (ishitasequeira) [reply pkg/pipelines/add_application.go]: resolved

File: pkg/manifest/init.go (hunk 11 above)
Comment (ishitasequeira) [reply pkg/manifest/init.go]: made the required changes

File: pkg/pipelines/add_application.go (hunk 13 above)
Comment (ishitasequeira) [reply pkg/pipelines/add_application.go]: resolved

File: pkg/pipelines/add_application.go (hunk 10 above)
Comment (ishitasequeira) [reply pkg/pipelines/add_application.go]: decided not to be incorporated in this PR.
